aws/request/http_request.go

```diff
@@ -9,20 +7,13 @@ import (
 )
 
 func copyHTTPRequest(r *http.Request, body io.ReadCloser) *http.Request {
-	req := &http.Request{
-		URL:           &url.URL{},
-		Header:        http.Header{},
-		Close:         r.Close,
-		Body:          body,
-		Host:          r.Host,
-		Method:        r.Method,
-		Proto:         r.Proto,
-		ContentLength: r.ContentLength,
-		// Cancel will be deprecated in 1.7 and will be replaced with Context
-		Cancel: r.Cancel,
-	}
-
+	req := new(http.Request)
+	*req = *r
```

Comment:
This was originally done as an explicit copy because of race conditions that could occur between the original request's `Body` and the underlying client's `Transport` could still have an outstanding call to `r.Body.Read`. The Go race checker flagged this as a race condition. This is why the individual fields were copied.

This is also why the Request wraps the original body reader in a [`offsetReader`](https://github.com/aws/aws-sdk-go/blob/master/aws/request/offset_reader.go). So that the underlying buffer can be rewound for the next retry request attempt.